pyiron_contrib/tinybase/container.py

```diff
@@ -0,0 +1,179 @@
+"""Generic Input Base Clases"""
+
+import abc
+import sys
+
+from pyiron_contrib.tinybase.storage import HasHDFAdapaterMixin
+
+from pyiron_base.interfaces.object import HasStorage
+from pyiron_atomistics.atomistics.structure.has_structure import HasStructure
+
+import numpy as np
+import matplotlib.pyplot as plt
+
+class StorageAttribute:
+    """
+    Create an attribute that is synced to a storage attribute.
+
+    It must be created on a class that derives from :class:`.HasStorage`.  It essentially behaves like a property that
+    writes and reads values to the underlying :attr:`.HasStorage.storage`. DataContainer of the class.  When accessing
+    this property before setting it, `None` is returned.
+
+    It's possible to modify the default value and the accepted type of values by using the builder-style :meth:`.type`
+    and :meth:`.default` methods.
+
+    >>> class MyType(HasStorage):
+    ...     a = StorageAttribute()
+    ...     b = StorageAttribute().type(int)
+    ...     c = StorageAttribute().default(list)
+    ...     d = StorageAttribute().default(lambda: 42).type(int)
+
+    >>> t = MyType()
+    >>> t.a # return None
+    >>> t.b = 3
+    >>> t.b
+    3
+    >>> t.b = 'asdf'
+    TypeError(f"{value} is not of type {self.value_type}")
+    >>> t.c
+    []
+    >>> t.d
+    42
+    """
+
+    def __init__(self):
+        self.name = None
+        self.value_type = object
+        self.default_constructor = None
+
+    def __set_name__(self, obj, name):
+        self.name = name
+
+    def __get__(self, obj, objtype=None):
+        try:
+            return obj.storage[self.name]
+        except KeyError:
+            if self.default_constructor is not None:
+                ret = obj.storage[self.name] = self.default_constructor()
+                return ret
+            else:
+                return None
+
+    def __set__(self, obj, value):
+        if isinstance(value, self.value_type):
```

Comment:
So, painfully, this won't work, at least not robustly.

The problem is that you can't use `isinstance` with `typing` generics. With a modern enough version of python you can get `isinstance("a", str | int)` and `isinstance("a", typing.Union[str, int])` both working, but there are other extremely useful and necessary type hints like `Literal` and `Callable` which throw an error, e.g. `isinstance("a", Literal["a"])`.

I wound up leaning on the `typeguard` package to handle both cases. It's doable, it's just more verbose than this check.

Reply:
Ah, value type is in this case really a concrete class not a type hint.  I don't see a fundamental problem with using `typeguard` or such to allow it, though.  More likely we'll also need range checks though, so I think at some point we'll just have to let devs provide functions to check values.  I think this is pretty similar to how we had talked about `traitlets`.  I'm not so worried that this might become a bit tricky, because we'll only need to define such a "general type check once and export it so that node devs can use it.